platform-sdk/swirlds-logging/src/main/java/com/swirlds/logging/api/internal/level/HandlerLoggingLevelConfig.java

```diff
@@ -0,0 +1,254 @@
+ * <p>
+ * {@code logging.level.com.swirlds=DEBUG}
+ * <p>
+ * In that case all packages and classes under the package {@code com.swirlds} will be configured to use the level
+ * DEBUG. If a more specific configuration is defined, that configuration will be used instead. For example:
+ * <p>
+ * {@code logging.level.com.swirlds.logging.api=INFO}
+ * <p>
+ * In that case all packages and classes under the package {@code com.swirlds.logging.api} will be configured to use the
+ * level INFO.
+ */
+public class HandlerLoggingLevelConfig {
+
+    private static final String PROPERTY_LOGGING_LEVEL = "logging.level";
+    private static final String PROPERTY_LOGGING_HANDLER_LEVEL = "logging.handler.%s.level";
+    private static final String PROPERTY_PACKAGE_LEVEL = "%s.";
+    private static final String PROPERTY_LOGGING_MARKER = "logging.marker";
+    private static final String PROPERTY_LOGGING_HANDLER_MARKER = "logging.handler.%s.marker";
+    private static final String PROPERTY_LOGGING_HANDLER_INHERIT_LEVELS = "logging.handler.%s.inheritLevels";
+    private static final ConfigLevel DEFAULT_LEVEL = ConfigLevel.INFO;
+
+    /**
+     * The emergency logger.
+     */
+    private static final EmergencyLogger EMERGENCY_LOGGER = EmergencyLoggerProvider.getEmergencyLogger();
+
+    /**
+     * The cache for the levels.
+     */
+    private final Map<String, ConfigLevel> levelCache;
+
+    /**
+     * The cache for the markers.
+     */
+    private final AtomicReference<Map<String, MarkerState>> markerConfigCache;
+
+    /**
+     * The configuration properties.
+     */
+    private final AtomicReference<Map<String, ConfigLevel>> levelConfigProperties;
+
+    /**
+     * The prefix for the configuration.
+     */
+    private final String name;
+
+    /**
+     * Creates a new logging level configuration based on the given configuration and handler name.
+     *
+     * @param configuration The configuration.
+     * @param name        The name.
+     */
+    public HandlerLoggingLevelConfig(@NonNull Configuration configuration, @Nullable String name) {
+        this.name = name;
+        this.levelCache = new ConcurrentHashMap<>();
+        this.markerConfigCache = new AtomicReference<>(new ConcurrentHashMap<>());
+        this.levelConfigProperties = new AtomicReference<>(new ConcurrentHashMap<>());
+        update(configuration);
+    }
+
+    /**
+     * Creates a new root configuration based on the given configuration.
+     * @param configuration The configuration.
+     */
+    public HandlerLoggingLevelConfig(@NonNull Configuration configuration) {
+        this(configuration, null);
+    }
+
+    /**
+     * Updates the configuration based on the given configuration. That method can be used to change the logging level
+     * dynamically at runtime.
+     *
+     * @param configuration The configuration.
+     */
+    public void update(final @NonNull Configuration configuration) {
```

Comment:
is the convention @NotNull final / final @NotNull or we don't care?

Reply:
Had the same question some time ago but missed to check it. We do not have it defined in our code style so far but it would make sense to define best practices (and maybe even have our code formatter check that). Can you create an issue for that?

Reply:
Created issue: https://github.com/hashgraph/hedera-services/issues/10380